src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -810,182 +759,195 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VariantMetadata,
-  }
+    val keyedRDD = kt.keyedRDD()
+      .filter { case (k, v) => k.toSeq.forall(_ != null) }
 
-  def nPartitions: Int = rdd.partitions.length
+    Option(vdsKey) match {
+      case Some(keys) =>
+        val keyEC = EvalContext(Map("v" -> (0, TVariant), "va" -> (1, vaSignature)))
+        val (vdsKeyType, vdsKeyFs) = keys.map(Parser.parseExpr(_, keyEC)).unzip
 
-  def annotateVariantsTable(path: String, variantExpr: String,
-    root: Option[String] = None, code: Option[String] = None,
-    config: TextTableConfiguration = TextTableConfiguration()): VariantSampleMatrix[T] = {
-    annotateVariantsTables(List(path), variantExpr, root, code, config)
-  }
+        if (!(keyTypes sameElements vdsKeyType))
+          fatal(
+            s"""method `annotateVariantsTable' encountered a mismatch between table keys and computed keys.
+               |  Computed keys:  [ ${ vdsKeyType.mkString(", ") } ]
+               |  Key table keys: [ ${ keyTypes.mkString(", ") } ]""".stripMargin)
 
-  def annotateVariantsTables(paths: Seq[String], variantExpr: String,
-    root: Option[String] = None, code: Option[String] = None,
-    config: TextTableConfiguration = TextTableConfiguration()): VariantSampleMatrix[T] = {
-    val files = hc.hadoopConf.globAll(paths)
-    if (files.isEmpty)
-      fatal("Arguments referred to no files")
-
-    val (struct, variantRDD) = TextTableReader.read(sparkContext)(files, config, nPartitions)
+        val thisRdd = rdd.map { case (v, (va, gs)) =>
+          keyEC.setAll(v, va)
+          (Row.fromSeq(vdsKeyFs.map(_ ())), v)
+        }
 
-    val (isCode, annotationExpr) = (root, code) match {
-      case (Some(r), None) => (false, r)
-      case (None, Some(c)) => (true, c)
-      case _ => fatal("this module requires one of `root' or 'code', but not both")
+        val joinedRDD = keyedRDD
+          .join(thisRdd)
+          .map { case (_, (table, v)) => (v, table: Annotation) }
+          .orderedRepartitionBy(rdd.orderedPartitioner)
+
+        annotateVariants(joinedRDD, finalType, inserter, product = product)
+
+      case None =>
+        keyTypes match {
+          case Array(TVariant) =>
+            val ord = keyedRDD
+              .map { case (k, v) => (k.getAs[Variant](0), v: Annotation) }
+              .toOrderedRDD(rdd.orderedPartitioner)
+
+            annotateVariants(ord, finalType, inserter, product = product)
+          case Array(TLocus) =>
+            import LocusImplicits.orderedKey
+            val ord = keyedRDD
+              .map { case (k, v) => (k.asInstanceOf[Row].get(0).asInstanceOf[Locus], v: Annotation) }
+              .toOrderedRDD(rdd.orderedPartitioner.mapMonotonic[Locus])
+
+            annotateLoci(ord, finalType, inserter, product = product)
+
+          case Array(TInterval) =>
+            val intRDD = keyedRDD
+              .map { case (k, v) => (k.getAs[Interval[Locus]](0), v: Annotation) }
+              .sortBy(_._1.start)
```

Comment:
You're right that it's cleaner. But we still can't do it, and here's why:

(using vds = gnomAD exome sites file)

filter_intervals for reference:
```
In [4]: %%timeit
   ...: vds.filter_intervals(Interval.parse('16:29.5m-30.2m')).count()
   ...:
hail: info: interval filter loaded 18 of 13592 partitions
[Stage 3:==================================================>      (16 + 2) / 18]hail: info: interval filter loaded 18 of 13592 partitions
[Stage 4:===============================================>         (15 + 3) / 18]hail: info: interval filter loaded 18 of 13592 partitions
[Stage 5:============================================>            (14 + 4) / 18]hail: info: interval filter loaded 18 of 13592 partitions
[Stage 6:============================================>            (14 + 4) / 18]1 loop, best of 3: 2.12 s per loop
```
old implementation
```
In [10]: %%timeit
    ...: vds.filter_variants_table(kt).count()
    ...:
hail: info: filtered to 18 of 13592 partitions
[Stage 13:====================================================>   (17 + 1) / 18]hail: info: filtered to 18 of 13592 partitions
[Stage 17:=================================================>      (16 + 2) / 18]hail: info: filtered to 18 of 13592 partitions
[Stage 21:====================================================>   (17 + 1) / 18]hail: info: filtered to 18 of 13592 partitions
[Stage 25:===========================================>            (14 + 4) / 18]1 loop, best of 3: 1.57 s per loop
```

shuffle implementation
```
In [4]: %%timeit
   ...: vds.filter_variants_table(kt).count()
   ...:
[Stage 9:==================================================>(13578 + 8) / 13592]1 loop, best of 3: 1min 52s per loop
```